### PR TITLE
fix: don't mix Vite plugins when spawning temporary Vite server

### LINKED ISSUE #6364: Running Vite server during build breaks plugin assumptions
### What version of `astro` are you using?

2.0.12+

### Are you using an SSR adapter? If so, which one?

N/A

### What package manager are you using?

pnpm

### What operating system are you using?

macOS

### Describe the Bug

In https://github.com/withastro/astro/pull/6238, we broke [`vite-plugin-pwa`](https://github.com/vite-pwa/vite-plugin-pwa/) (and possibly others) because we started running a temporary Vite server during the build to load Content Collections. `vite-plugin-pwa` uses [`apply: 'serve'`](https://github.com/vite-pwa/vite-plugin-pwa/blob/main/src/plugins/dev.ts#L32) as a heuristic for running the dev server, which no longer works when we start a temporary server _and pass all plugins_ to that server.

cc @userquin 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-kyecnu?file=integration.js,astro.config.mjs,package.json

### Participation

- [X] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by userquin**
## Changes

- Closes #6364
- Supersedes #6366
- Filter plugins using the Astro command
- /cc @bholmesdev
- /cc @natemoo-re

## Testing

Added simple test with disparate plugins for static build.

## Docs

Bug fix only: maybe we need to add some hints for integrations registering plugins, check for example `@vite-pwa/astro` integration logic: https://github.com/vite-pwa/astro/blob/main/src/index.ts#L31

### CODE REVIEW

**Comment by userquin** [comment]
We need the command, not the mode.

**Comment by userquin** [comment]
for preview, integrations should take care, and for vite user plugins we should also filter them before starting Vite preview

File: packages/astro/src/core/create-vite.ts [hunk 1]
```diff
@@ -170,7 +172,31 @@ export async function createVite(
 	//   3. integration-provided vite config, via the `config:setup` hook
 	//   4. command vite config, passed as the argument to this function
 	let result = commonConfig;
-	result = vite.mergeConfig(result, settings.config.vite || {});
+	// command will be undefined when running sync
+	if (command && command !== 'preview') {
+		let plugins = settings.config.vite?.plugins;
+		if (plugins) {
+			const { plugins: _, ...rest } = settings.config.vite
+			const applyToFilter = mode === 'build' ? 'serve' : 'build'
```
Comment (userquin) [inline packages/astro/src/core/create-vite.ts:180]: this should be `const applyToFilter = command === 'build' ? 'serve' : 'build'`

Comment (userquin) [reply packages/astro/src/core/create-vite.ts:180]: we can `build` in `dev` mode

File: packages/astro/src/core/create-vite.ts [hunk 2]
```diff
@@ -30,6 +30,8 @@ interface CreateViteOptions {
 	settings: AstroSettings;
 	logging: LogOptions;
 	mode: 'dev' | 'build' | string;
+	// will be undefined when using `sync`
+	command?: 'dev' | 'build' | 'preview';
```
Comment (userquin) [inline packages/astro/src/core/create-vite.ts:34]: remove `'preview'`, Astro will use Vite `preview`

File: packages/astro/src/core/create-vite.ts [hunk 3]
```diff
@@ -170,7 +172,31 @@ export async function createVite(
 	//   3. integration-provided vite config, via the `config:setup` hook
 	//   4. command vite config, passed as the argument to this function
 	let result = commonConfig;
-	result = vite.mergeConfig(result, settings.config.vite || {});
+	// command will be undefined when running sync
+	if (command && command !== 'preview') {
```
Comment (userquin) [inline packages/astro/src/core/create-vite.ts:176]: onde remoed `'preview'` from `CreateViteOptions` changethis to `if (command) {`

**Comment by userquin** [comment]
don't merge yet, preparing test

**Comment by userquin** [comment]
We've another problem, when a plugin is included it can be excluded on sync call since we don't have command/action when using apply with function, we need to review it.

File: packages/astro/src/core/create-vite.ts [hunk 4]
```diff
@@ -170,7 +172,31 @@ export async function createVite(
...
+				}
+
+				return true;
+			});
+
+			result = vite.mergeConfig(result, { ...rest, plugins });
+		} else {
+			result = vite.mergeConfig(result, settings.config.vite || {});
+		}
+	}
```
Comment (userquin) [inline packages/astro/src/core/create-vite.ts:199]: add else case without plugins

**Review by natemoo-re — APPROVED** [review]
Great fix and awesome test! 

Test failure is unrelated.

File: packages/astro/src/core/create-vite.ts [hunk 5]
```diff
@@ -170,7 +172,31 @@ export async function createVite(
...
 	let result = commonConfig;
-	result = vite.mergeConfig(result, settings.config.vite || {});
+	// command will be undefined when running sync
+	if (command) {
+		let plugins = settings.config.vite?.plugins;
+		if (plugins) {
+			const { plugins: _, ...rest } = settings.config.vite
+			const applyToFilter = command === 'build' ? 'serve' : 'build'
+			const applyArgs = [{...settings.config.vite, mode}, { command, mode }]
+			// @ts-ignore we know what are doing
```
Comment (ematipico) [inline packages/astro/src/core/create-vite.ts:182]: Please use `@ts-expect-error`, so we can remove the comment if the compiler gets better.

Would mind changing the comment too? Just saying "we know what we are doing" doesn't seem useful to other developers. Can we explain _what_ we are doing?

File: packages/astro/src/core/create-vite.ts [hunk 6]
```diff
@@ -170,7 +172,31 @@ export async function createVite(
 	//   3. integration-provided vite config, via the `config:setup` hook
 	//   4. command vite config, passed as the argument to this function
 	let result = commonConfig;
-	result = vite.mergeConfig(result, settings.config.vite || {});
+	// command will be undefined when running sync
+	if (command) {
```
Comment (ematipico) [inline packages/astro/src/core/create-vite.ts:176]: I'd like to see a comment that explains what this new code does.

**Review by ematipico — COMMENTED** [review]
I'm not against the change, but I'd like to see better comments

File: packages/astro/src/core/create-vite.ts (hunk 5 above)
Comment (userquin) [reply packages/astro/src/core/create-vite.ts:182]: imagen

Comment (userquin) [reply packages/astro/src/core/create-vite.ts:182]: we need to traverse flat plugins array to filter them

**Review by ematipico — APPROVED** [review]

**Review by bholmesdev — COMMENTED** [review]
Added some suggestions too clear up in-line comments. Great fix though, was unaware Vite plugins have an `apply` utility!

File: packages/webapi/mod.d.ts [hunk 7]
```diff
@@ -1,13 +1,13 @@
 // organize-imports-ignore
```
[the remark this answers is not in the document]
Comment (natemoo-re) [reply packages/webapi/mod.d.ts:1]: I can't figure out why this is here. Git seems to think the file is up-to-date with `main`. Resolving so we can merge.

**Pull request merged**